### PR TITLE
HDDS-15170. Add mock-based unit tests for DataStream write path

### PR DESCRIPTION
**Pull request by fapifta**
## What changes were proposed in this pull request?

Add `MockDatanodePipeline` test harness that simulates a Ratis DataStream pipeline using mocked `XceiverClientRatis` and a concrete `DataStreamOutput` implementation, with failure injection for chunk writes, putBlock, and watchForCommit
Add `TestBlockDataStreamOutput` with 12 block-level tests covering sub-chunk writes, exact chunk boundaries, flush boundaries, back-pressure via stream window, hsync happy/error paths, close behavior, write-after-close rejection, ack tracking, data integrity, and putBlock metadata
Add `TestKeyDataStreamOutput` with 7 key-level tests covering write-and-commit, cross-block boundary allocation, hsync-to-OM integration, container-close retry, multiple hsyncs, and write-after-close rejection                                                                                                                       

Two block-level tests (`hsyncPropagatesIOException`, `hsyncPropagatesWatchFailure`) and one key-level test (`hsyncWithBlockErrorDoesNotCallOmHsync`) are currently disabled — they document a known hsync exception handling issue where `BlockDataStreamOutput.hsync()` silently swallows all exceptions via an empty catch block, and `KeyDataStreamOutput.handleFlushOrClose()` retries hsync failures as if they were write failures. These tests will be enabled when the fix lands.

Generated-by: Claude - Opus 4.6

## What is the link to the Apache JIRA
https://issues.apache.org/jira/browse/HDDS-15170

## How was this patch tested?
The added unit tests succeed.

### CODE REVIEW

File: hadoop-hdds/client/dev-support/findbugsExcludeFile.xml [hunk 1]
```diff
@@ -19,4 +19,13 @@
     <Class name="org.apache.hadoop.hdds.scm.storage.ByteArrayReader"></Class>
     <Bug pattern="EI_EXPOSE_REP2" /> <!-- "Deep copy byte[] has bad impact on performance" -->
   </Match>
+  <Match>
+    <!--
+    Required until we have https://github.com/spotbugs/spotbugs/pull/3410 (SpotBugs 4.8)
+    FakeDatanodePipeline Line "doReturn(pipeline).when(xceiverClient).getPipeline();" (115 at time of this addition
+    is reported but is a false positive.
```
Comment (adoroszlai) [inline hadoop-hdds/client/dev-support/findbugsExcludeFile.xml]: The statement can be replaced with:

```
    when(xceiverClient.getPipeline()).thenReturn(pipeline);
```

to avoid the need for findbugs exclusion.

See also HDDS-12472.

Comment (fapifta) [reply hadoop-hdds/client/dev-support/findbugsExcludeFile.xml]: You are right, I wanted to preserve the same form of mock definition there, but all in all it is also ok if we mix it.
Thank you for taking a look!

**Comment by adoroszlai** [comment]
@fapifta Can you please re-run _all_ checks in your fork https://github.com/fapifta/hadoop-ozone/actions/runs/25658037221?  I'd like to compare code coverage before/after the patch.  (Unfortunately re-run failed acceptance check works only within 1 day of failure due to artifact expiry.)

**Comment by fapifta** [comment]
@adoroszlai I started it just now. https://github.com/fapifta/hadoop-ozone/actions/runs/25658037221

**Comment by adoroszlai** [comment]
This PR increases coverage of `BlockDataStreamOutput` from 77% to 81%.

**Comment by fapifta** [comment]
@peterxcli are you planning to review this PR?

It would be nice to merge this one.

**Comment by peterxcli** [comment]
> @peterxcli are you planning to review this PR?
> 
> It would be nice to merge this one.

sure, please give me some time!

File: hadoop-hdds/client/src/test/java/org/apache/hadoop/hdds/scm/storage/TestBlockDataStreamOutput.java [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+        fake.getClientFactory(),
+        fake.getPipeline(),
+        createConfig(),
+        null,  // no token
+        bufferList);
+  }
+
+  @Test
+  void writeSubChunkThenClose() throws Exception {
+    FakeDatanodePipeline fake = new FakeDatanodePipeline();
```
Comment (peterxcli) [inline hadoop-hdds/client/src/test/java/org/apache/hadoop/hdds/scm/storage/TestBlockDataStreamOutput.java]: nit
```suggestion
    FakeDatanodePipeline pipeline = new FakeDatanodePipeline();
```

File: hadoop-hdds/client/src/test/java/org/apache/hadoop/hdds/scm/storage/FakeDatanodePipeline.java [hunk 3]
```diff
@@ -0,0 +1,372 @@
...
+    @Override
+    public long getStreamOffset() {
+      return 0;
+    }
+
+    @Override
+    public long getDataLength() {
+      return bytesWritten;
+    }
+  }
```
Comment (peterxcli) [inline hadoop-hdds/client/src/test/java/org/apache/hadoop/hdds/scm/storage/FakeDatanodePipeline.java]:
```suggestion
  private static DataStreamReply dataStreamReply(long bytesWritten) {
    DataStreamReply reply = mock(DataStreamReply.class);
    when(reply.isSuccess()).thenReturn(true);
    when(reply.getBytesWritten()).thenReturn(bytesWritten);
    when(reply.getDataLength()).thenReturn(bytesWritten);
    when(reply.getCommitInfos()).thenReturn(Collections.emptyList());
    return reply;
  }
```

File: hadoop-hdds/client/src/test/java/org/apache/hadoop/hdds/scm/storage/FakeDatanodePipeline.java [hunk 4]
```diff
@@ -0,0 +1,372 @@
...
+
+    @Override
+    public long getStreamOffset() {
+      return 0;
+    }
+
+    @Override
+    public long getDataLength() {
+      return bytesWritten;
+    }
```
Comment (peterxcli) [inline hadoop-hdds/client/src/test/java/org/apache/hadoop/hdds/scm/storage/FakeDatanodePipeline.java]: we can move this to class init with mock invocation just like `putblock` and `watchForCommit`, and we can avoid adding derived class lines.

```suggestion
  doAnswer(invocation -> {
      ByteBuffer src = invocation.getArgument(0);
      Iterable<WriteOption> options = invocation.getArgument(1);
      int size = src.remaining();
      for (WriteOption option : options) {
        if (option == StandardWriteOption.CLOSE) {
          if (!receivedChunks.isEmpty()) {
            receivedChunks.remove(receivedChunks.size() - 1);
          }
          src.position(src.limit());
          return CompletableFuture.completedFuture(dataStreamReply(size));
        }
      }

      int count = chunkCount.incrementAndGet();
      if (count > chunkFailAfter && chunkFailure != null) {
        CompletableFuture<DataStreamReply> failed = new CompletableFuture<>();
        failed.completeExceptionally(chunkFailure.get());
        return failed;
      }

      byte[] data = new byte[size];
      src.get(data);
      receivedChunks.add(data);
      return CompletableFuture.completedFuture(dataStreamReply(data.length));
    }).when(fakeOut)
        .writeAsync(any(ByteBuffer.class), any(Iterable.class));
```

File: hadoop-hdds/client/src/test/java/org/apache/hadoop/hdds/scm/storage/MockDatanodePipeline.java [hunk 5]
```diff
@@ -0,0 +1,372 @@
...
+    this.putBlockFailAfter = n;
+    this.putBlockFailure = err;
+    return this;
+  }
+
+  public FakeDatanodePipeline failWatchAfter(int n, Supplier<Throwable> err) {
+    this.watchFailAfter = n;
+    this.watchFailure = err;
+    return this;
+  }
```
Comment (peterxcli) [inline hadoop-hdds/client/src/test/java/org/apache/hadoop/hdds/scm/storage/MockDatanodePipeline.java:262]: add a builder for fake datanode pipeline?

File: hadoop-hdds/client/src/test/java/org/apache/hadoop/hdds/scm/storage/FakeDatanodePipeline.java [hunk 6]
```diff
@@ -0,0 +1,372 @@
...
+
+/**
+ * A stateful test harness that simulates a datanode pipeline for {@link BlockDataStreamOutput} unit tests.
+ * Replaces a real Ratis pipeline with mocked {@link XceiverClientRatis} and a concrete {@link DataStreamOutput}
+ * implementation.
+ *
+ * <p>Tracks all chunks written, putBlock calls, and watchForCommit calls.
+ * Configurable failure injection for each operation.
+ */
+public class FakeDatanodePipeline {
```
Comment (peterxcli) [inline hadoop-hdds/client/src/test/java/org/apache/hadoop/hdds/scm/storage/FakeDatanodePipeline.java]: nit:
```suggestion
public class MockDataStreamPipeline {
```

so we can distinguished this v2 pipeline from v1 write chunk + put block based pipeline, which should have its own `MockBlockPipeline`.

File: hadoop-hdds/client/src/test/java/org/apache/hadoop/hdds/scm/storage/TestBlockDataStreamOutput.java [hunk 7]
```diff
@@ -0,0 +1,267 @@
...
+    // Window = 500B = 5 chunks. Writing 500B should trigger back-pressure.
+    byte[] data = randomBytes(500);
+    try (BlockDataStreamOutput stream = createStream(fake)) {
+      stream.write(ByteBuffer.wrap(data), 0, data.length);
+      // 5 chunks written, putBlock at 400B boundary, back-pressure at 500B
+      // should have triggered watchForCommit
+      assertThat(fake.getWatchForCommitCount())
+          .as("watchForCommit should be called for back-pressure")
+          .isGreaterThanOrEqualTo(1);
```
Comment (peterxcli) [inline hadoop-hdds/client/src/test/java/org/apache/hadoop/hdds/scm/storage/TestBlockDataStreamOutput.java]:
```suggestion
          .isEqualTo(1);
```

File: hadoop-hdds/client/src/test/java/org/apache/hadoop/hdds/scm/storage/TestBlockDataStreamOutput.java [hunk 8]
```diff
@@ -0,0 +1,267 @@
...
+      // If the async error hasn't been checked yet, close will catch it
+      stream.close();
+      // If we get here without exception, the async error will surface later
+    } catch (IOException e) {
+      // Expected — either write or close threw due to chunk failure
+      assertThat(e.getMessage()).contains("chunk write failed");
+      return;
+    }
+    // If neither threw, the error was silently lost (also a problem,
+    // but let's not fail the test for this secondary issue)
```
Comment (peterxcli) [inline hadoop-hdds/client/src/test/java/org/apache/hadoop/hdds/scm/storage/TestBlockDataStreamOutput.java]: let's make the exception capture scope to be more precise.
```suggestion
    stream.write(ByteBuffer.wrap(data2), 0, data2.length);
    IOException e = assertThrows(OMException.class, () -> stream.close());
    assertThat(e.getMessage()).contains("chunk write failed");
```

File: hadoop-hdds/client/src/test/java/org/apache/hadoop/hdds/scm/storage/TestBlockDataStreamOutput.java [hunk 9]
```diff
@@ -0,0 +1,267 @@
...
+  void writeAfterCloseThrows() throws Exception {
+    FakeDatanodePipeline fake = new FakeDatanodePipeline();
+    BlockDataStreamOutput stream = createStream(fake);
+    byte[] data = randomBytes(CHUNK_SIZE);
+    stream.write(ByteBuffer.wrap(data), 0, data.length);
+    stream.close();
+
+    assertThrows(IOException.class,
+        () -> stream.write(ByteBuffer.wrap(data), 0, data.length),
+        "write() after close() should throw IOException");
```
Comment (peterxcli) [inline hadoop-hdds/client/src/test/java/org/apache/hadoop/hdds/scm/storage/TestBlockDataStreamOutput.java]: assert exception message equal to something?
```suggestion
    assertThrows(IOException.class,
        () -> stream.write(ByteBuffer.wrap(data), 0, data.length));
    // TODO: assert exception message equal to something
```

File: hadoop-hdds/client/src/test/java/org/apache/hadoop/hdds/scm/storage/TestBlockDataStreamOutput.java [hunk 10]
```diff
@@ -0,0 +1,267 @@
...
+  void hsyncFlushesAndWaitsForCommit() throws Exception {
+    FakeDatanodePipeline fake = new FakeDatanodePipeline();
+    try (BlockDataStreamOutput stream = createStream(fake)) {
+      byte[] data = randomBytes(200);
+      stream.write(ByteBuffer.wrap(data), 0, data.length);
+      stream.hsync();
+      // 2 chunks, 1 putBlock (from hsync), watch called
+      assertEquals(2, fake.getReceivedChunks().size());
+      assertThat(fake.getReceivedPutBlocks()).isNotEmpty();
+      assertThat(fake.getWatchForCommitCount()).isGreaterThanOrEqualTo(1);
```
Comment (peterxcli) [inline hadoop-hdds/client/src/test/java/org/apache/hadoop/hdds/scm/storage/TestBlockDataStreamOutput.java]: let's make the assertion deterministic.
```suggestion
      assertEquals(1, fake.getReceivedPutBlocks().size());
      assertEquals(1, fake.getWatchForCommitCount());
```

File: hadoop-hdds/client/src/test/java/org/apache/hadoop/hdds/scm/storage/TestBlockDataStreamOutput.java [hunk 11]
```diff
@@ -0,0 +1,267 @@
...
+    FakeDatanodePipeline fake = new FakeDatanodePipeline();
+    // Fail the first putBlock
+    fake.failPutBlockAfter(0, () -> new IOException("simulated putBlock fail"));
+
+    BlockDataStreamOutput stream = createStream(fake);
+    byte[] data = randomBytes(200);
+    stream.write(ByteBuffer.wrap(data), 0, data.length);
+
+    // hsync should propagate the IOException from the failed putBlock
+    assertThrows(IOException.class, stream::hsync, "hsync() must propagate IOException from failed putBlock");
```
Comment (peterxcli) [inline hadoop-hdds/client/src/test/java/org/apache/hadoop/hdds/scm/storage/TestBlockDataStreamOutput.java:153]: so should we expose exception from the hysnc?

File: hadoop-hdds/client/src/test/java/org/apache/hadoop/hdds/scm/storage/TestBlockDataStreamOutput.java [hunk 12]
```diff
@@ -0,0 +1,267 @@
...
+    long totalLength = 0;
+    for (ContainerProtos.ContainerCommandRequestProto pb : fake.getReceivedPutBlocks()) {
+      ContainerProtos.BlockData blockData = pb.getPutBlock().getBlockData();
+      for (ContainerProtos.ChunkInfo chunk : blockData.getChunksList()) {
+        totalChunks++;
+        totalLength += chunk.getLen();
+        // Chunk names should be sequential
+        assertThat(chunk.getChunkName()).contains("_chunk_");
+      }
+    }
```
Comment (peterxcli) [inline hadoop-hdds/client/src/test/java/org/apache/hadoop/hdds/scm/storage/TestBlockDataStreamOutput.java]:
```suggestion
    assertEquals(1, pipeline.getReceivedPutBlocks().size());

	ContainerProtos.BlockData blockData = pipeline.getReceivedPutBlocks()
	    .get(0)
	    .getPutBlock()
	    .getBlockData();
	
	long totalLength = 0;
	for (ContainerProtos.ChunkInfo chunk : blockData.getChunksList()) {
	  totalLength += chunk.getLen();
	  assertThat(chunk.getChunkName()).contains("_chunk_");
	}
```

**Review by peterxcli — COMMENTED** [review]
thanks @fapifta for introducing the mock pipeline, so we can finally do unit test on data stream output.

left some comments, will review key data stream output test later.

**Comment by fapifta** [comment]
Hi @peterxcli thank you for your thorough review on these test additions. The Fake, now MockDatanodePipeline related suggestions were really great.
I tried to address the review comments even in places where they were not directly added, but a similar comment matched onto the situation elsewhere in the code also. (Like rename to pipeline instead of fake, and explicitly check for the number of operations done.)
I also removed the atomic key creation related test, as that is moved from the stream to S3 in an other change that affected the test.
I found a flakiness with the injected failure and how it surfaces I addressed that also.

I resolved all the relevant comments, except one that I would like to address directly there before we resolve, if you accept my answer there please resolve the review.
Please also re-check the suggested changes once you have the time.

File: hadoop-hdds/client/src/test/java/org/apache/hadoop/hdds/scm/storage/MockDatanodePipeline.java (hunk 5 above)
Comment (fapifta) [reply hadoop-hdds/client/src/test/java/org/apache/hadoop/hdds/scm/storage/MockDatanodePipeline.java:262]: As currently its constructor has only one parameter, I would consider that as boilerplate bloat... Do you think it is necessary, or is it enough to add a builder if the number of constructor parameters increase further?

File: hadoop-hdds/client/src/test/java/org/apache/hadoop/hdds/scm/storage/TestBlockDataStreamOutput.java (hunk 11 above)
Comment (fapifta) [reply hadoop-hdds/client/src/test/java/org/apache/hadoop/hdds/scm/storage/TestBlockDataStreamOutput.java:153]: Yes, that is what is being prepared here, but I wanted to keep things clean, and just added the tests that surfaces the problem that the exception is not exposed.
The tests should be unskipped in a follow up commit that surfaces the exception from hsync itslef, and enable these two tests. These two tests fail now.

The rationale is that if the exception is not exposed, then a client might falsely think that the sync was successful, while it failed, so the client has a different view on what was persisted and what was not.

Comment (peterxcli) [reply hadoop-hdds/client/src/test/java/org/apache/hadoop/hdds/scm/storage/TestBlockDataStreamOutput.java:153]: > if the exception is not exposed, then a client might falsely think that the sync was successful, while it failed, so the client has a different view on what was persisted and what was not.

make sense, is there any ticket for this?

File: hadoop-hdds/client/src/test/java/org/apache/hadoop/hdds/scm/storage/MockDatanodePipeline.java (hunk 5 above)
Comment (peterxcli) [reply hadoop-hdds/client/src/test/java/org/apache/hadoop/hdds/scm/storage/MockDatanodePipeline.java:262]: I'm ok with the current version, but I have the following structure in my mind:
```java
pipeline = MockDatanodePipeline.newBuilder()
  .setFailPutBlockAfter(...)
  .setFailWatchAfter(...)
  .setFailChunkAfter(...)

// or
pipeline = MockDatanodePipeline.newBuilder()
  .setFailChunkAfter(...)

// or
pipeline = MockDatanodePipeline.newBuilder()
  .setFailWatchAfter(...)
  .setFailChunkAfter(...)
```

Comment (peterxcli) [reply hadoop-hdds/client/src/test/java/org/apache/hadoop/hdds/scm/storage/MockDatanodePipeline.java:262]: but looks like there is no multiple failure injection now, so maybe the builder is unnecessary for now.

**Review by peterxcli — APPROVED** [review]
Thanks @fapifta for the update, patch now LGTM. 

btw I've replied two of your comments, please also take a look, Thanks!
btw, please update the pr description corespondingly if possible, Thanks!

File: hadoop-hdds/client/src/test/java/org/apache/hadoop/hdds/scm/storage/TestBlockDataStreamOutput.java (hunk 11 above)
Comment (fapifta) [reply hadoop-hdds/client/src/test/java/org/apache/hadoop/hdds/scm/storage/TestBlockDataStreamOutput.java:153]: I created [HDDS-15169](https://issues.apache.org/jira/browse/HDDS-15169) to post a series of changes, the hsync exception exposure comes in the second defined task, though I have not created all of them, it was in flux back then when I created the first two.
[HDDS-15225](https://issues.apache.org/jira/browse/HDDS-15225) is the hsync change.

File: hadoop-hdds/client/src/test/java/org/apache/hadoop/hdds/scm/storage/MockDatanodePipeline.java (hunk 5 above)
Comment (fapifta) [reply hadoop-hdds/client/src/test/java/org/apache/hadoop/hdds/scm/storage/MockDatanodePipeline.java:262]: I just put this aside yesterday as I wanted to explore the idea, but I realized that the failure setup is a fluid API, so you can do:
```
MockDatanodePipeline pipeline = new MockDatanodePipeline(blockId)
  .setFailPutBlockAfter(...)
  .setFailChunkAfter(...);
```

**Comment by fapifta** [comment]
Thank you for the reviews @adoroszlai and @peterxcli!
